Add quantile compression codec for floating point data types

rgolovachev opened the pull request:
### Changelog category (leave one):
- New Feature


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added [quantile compression codec](https://github.com/mwlon/pcodec), which is currently called "pcodec".

Requested here:
[#58394](https://github.com/ClickHouse/ClickHouse/issues/58394)

rgolovachev committed:
Add quantile compression codec for floating point data types

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=62823) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=62823) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=62823) it.</sub>

rgolovachev committed:
Resolved conflicts

rgolovachev committed:
pco rust package moved

alexey-milovidov commented:
How does it fare in performance and compression ratio?

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #62823 from rgolovachev/ClickHouse:add-quantile-codec -->
*This is an automated comment for commit 7a6433b4fb000eaa7c015f4e48ca25f4ae18ce5f with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62823/7a6433b4fb000eaa7c015f4e48ca25f4ae18ce5f/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

Felixoid commented:
@Algunenano Roman asked for support adding the library in the telegram devs chat

Algunenano commented:
The steps are pretty simple, but you'll need to make sure you are generating cmake files from scratch after following them.


Example:

### Checking the current status

```
$ git fetch blessed pull/62823/head:add-quantile-codec
$ git checkout add-quantile-codec
$ git submodule sync && git submodule update --init --recursive -f
$ mkdir build_new_rust
$ cd build_new_rust
$ cmake ..
$ ninja rust/all
[0/2] Re-checking globbed directories...
[0/7] cd /mnt/ch/ClickHouse/build_new_rust/rust/workspace/skim && /...profile=release --offline --locked -- -Cdefault-linker-libraries=no
error: no matching package named `pco` found
location searched: registry `crates-io`
required by package `_ch_rust_pco v0.1.0 (/mnt/ch/ClickHouse/build_new_rust/rust/workspace/pco)`
As a reminder, you're using offline mode (--offline) which can sometimes cause surprising resolution failures, if this error is too confusing you may wish to retry without the offline flag.
[1/7] cd /mnt/ch/ClickHouse/rust && /usr/bin/cmake -DFROM=workspace...e/build_new_rust/rust -P /mnt/ch/ClickHouse/rust/copy_exclude.cmake
FAILED: rust/workspace/CMakeFiles/_cargo-build__ch_rust_skim_rust /mnt/ch/ClickHouse/build_new_rust/rust/workspace/CMakeFiles/_cargo-build__ch_rust_skim_rust 
cd /mnt/ch/ClickHouse/build_new_rust/rust/workspace/skim && /usr/bin/cmake -E env CARGO_TARGET_X86_64_UNKNOWN_LINUX_GNU_LINKER=/usr/lib/ccache/bin/clang CC_x86_64_unknown_linux_gnu=/usr/lib/ccache/bin/clang CXX_x86_64_unknown_linux_gnu=/usr/lib/ccache/bin/clang++ AR_x86_64_unknown_linux_gnu=/usr/bin/llvm-ar CORROSION_BUILD_DIR=/mnt/ch/ClickHouse/build_new_rust/rust/workspace CARGO_BUILD_RUSTC=/home/raul/.rustup/toolchains/nightly-2024-04-01-x86_64-unknown-linux-gnu/bin/rustc /home/raul/.rustup/toolchains/nightly-2024-04-01-x86_64-unknown-linux-gnu/bin/cargo rustc --lib --target=x86_64-unknown-linux-gnu --package _ch_rust_skim_rust --manifest-path /mnt/ch/ClickHouse/build_new_rust/rust/workspace/skim/Cargo.toml --target-dir /mnt/ch/ClickHouse/build_new_rust/./cargo/build --profile=release --offline --locked -- -Cdefault-linker-libraries=no
ninja: build stopped: subcommand failed.
```

#### Check the rust toolchain

```bash
$ cd ../rust
$ rustup toolchain list
nightly-2023-07-04-x86_64-unknown-linux-gnu
nightly-2024-04-01-x86_64-unknown-linux-gnu (default)
```


#### Follow the guide

```
export CH_TOP_DIR=$(git rev-parse --show-toplevel)
export RUSTC_ROOT=$(rustc --print=sysroot)

cd "$CH_TOP_DIR"/rust/workspace

cargo local-registry --git --sync Cargo.lock "$CH_TOP_DIR"/contrib/rust_vendor
cp "$RUSTC_ROOT"/lib/rustlib/src/rust/Cargo.lock "$RUSTC_ROOT"/lib/rustlib/src/rust/library/std/
cargo local-registry --no-delete --git --sync "$RUSTC_ROOT"/lib/rustlib/src/rust/library/std/Cargo.lock "$CH_TOP_DIR"/contrib/rust_vendor
cp "$RUSTC_ROOT"/lib/rustlib/src/rust/Cargo.lock "$RUSTC_ROOT"/lib/rustlib/src/rust/library/test/
cargo local-registry --no-delete --git --sync "$RUSTC_ROOT"/lib/rustlib/src/rust/library/test/Cargo.lock "$CH_TOP_DIR"/contrib/rust_vendor

cargo vendor --no-delete --locked "$CH_TOP_DIR"/contrib/rust_vendor
cd "$RUSTC_ROOT"/lib/rustlib/src/rust/library/std/
cargo vendor --no-delete "$CH_TOP_DIR"/contrib/rust_vendor
cd "$RUSTC_ROOT"/lib/rustlib/src/rust/library/test/
cargo vendor --no-delete "$CH_TOP_DIR"/contrib/rust_vendor

rm -rf "$CH_TOP_DIR"/contrib/rust_vendor/winapi* "$CH_TOP_DIR"/contrib/rust_vendor/windows*

rm "$RUSTC_ROOT"/lib/rustlib/src/rust/library/std/Cargo.lock "$RUSTC_ROOT"/lib/rustlib/src/rust/library/test/Cargo.lock
cd "$CH_TOP_DIR"/rust/workspace
```

#### Check the result

The previous step should have added the new packages to the rust_vendor subdmoule

```
$ cd ../../contrib/rust_vendor
$ git status
HEAD detached at 08e82ca
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        better_io-0.1.0.crate
        better_io/
        index/3/p/pco
        index/be/tt/
        index/ra/nd/rand_xoshiro
        pco-0.2.3.crate
        pco/
        rand_xoshiro-0.6.0.crate
        rand_xoshiro/

nothing added to commit but untracked files present (use "git add" to track)
```


#### Build again

Note that I'm removing all the generated cmake files:

```
$ cd ../../build_new_rust/
$ rm -rf *
$ cmake ..
$ ninja rust/all
[0/2] Re-checking globbed directories...
[0/7] cd /mnt/ch/ClickHouse/build_new_rust/rust/workspace/skim && /...profile=release --offline --locked -- -Cdefault-linker-libraries=no
   Compiling proc-macro2 v1.0.79
   Compiling unicode-ident v1.0.12
   Compiling crossbeam-utils v0.8.19
   Compiling autocfg v1.2.0
   Compiling cc v1.0.92
   Compiling libc v0.2.153
   Compiling syn v1.0.109
   Compiling cfg-if v1.0.0
   Compiling fnv v1.0.7
   Compiling strsim v0.10.0
   Compiling ident_case v1.0.1
   Compiling scratch v1.0.7
   Compiling once_cell v1.19.0
   Compiling rayon-core v1.12.1
   Compiling termcolor v1.4.1
   Compiling cxxbridge-flags v1.0.121
   Compiling unicode-width v0.1.11
   Compiling memchr v2.7.2
   Compiling bitflags v1.3.2
   Compiling powerfmt v0.2.0
   Compiling iana-time-zone v0.1.60
   Compiling regex-syntax v0.8.3
   Compiling lazy_static v1.4.0
   Compiling num-conv v0.1.0
   Compiling time-core v0.1.2
   Compiling log v0.4.21
   Compiling arrayvec v0.7.4
   Compiling utf8parse v0.2.1
   Compiling either v1.10.0
   Compiling pin-utils v0.1.0
   Compiling beef v0.5.2
   Compiling thread_local v1.1.8
   Compiling codespan-reporting v0.11.1
   Compiling num-traits v0.2.18
   Compiling memoffset v0.6.5
   Compiling deranged v0.3.11
   Compiling fuzzy-matcher v0.3.7
   Compiling aho-corasick v1.1.3
   Compiling quote v1.0.35
   Compiling crossbeam-epoch v0.9.18
   Compiling crossbeam-channel v0.5.12
   Compiling crossbeam-queue v0.3.11
   Compiling syn v2.0.58
   Compiling vte_generate_state_changes v0.1.1
   Compiling link-cplusplus v1.0.9
   Compiling cxx v1.0.121
   Compiling crossbeam-deque v0.8.5
   Compiling time v0.3.34
   Compiling vte v0.11.1
   Compiling crossbeam v0.8.4
   Compiling defer-drop v1.3.0
   Compiling dirs-sys-next v0.1.2
   Compiling nix v0.24.3
   Compiling nix v0.25.1
   Compiling dirs-next v2.0.0
   Compiling term v0.7.0
   Compiling chrono v0.4.37
   Compiling rayon v1.10.0
   Compiling regex-automata v0.4.6
   Compiling timer v0.2.0
   Compiling tuikit v0.5.0 (https://github.com/azat-rust/tuikit.git?rev=e1994c0e03ff02c49cf1471f0cc3cbf185ce0104#e1994c0e)
   Compiling cxx-build v1.0.121
   Compiling cxxbridge-macro v1.0.121
   Compiling darling_core v0.14.4
   Compiling regex v1.10.4
   Compiling darling_macro v0.14.4
   Compiling _ch_rust_skim_rust v0.1.0 (/mnt/ch/ClickHouse/build_new_rust/rust/workspace/skim)
   Compiling darling v0.14.4
   Compiling derive_builder_core v0.11.2
    Compiling derive_builder_macro v0.11.2
   Compiling derive_builder v0.11.2
   Compiling skim v0.10.4
    Finished `release` profile [optimized] target(s) in 4.64s
[2/7] cd /mnt/ch/ClickHouse/build_new_rust/rust/workspace/prql && /...profile=release --offline --locked -- -Cdefault-linker-libraries=no
   Compiling proc-macro2 v1.0.79
   Compiling serde v1.0.197
   Compiling libc v0.2.153
   Compiling version_check v0.9.4
   Compiling rustversion v1.0.15
   Compiling zerocopy v0.7.32
   Compiling allocator-api2 v0.2.16
   Compiling adler v1.0.2
   Compiling gimli v0.28.1
   Compiling heck v0.4.1
   Compiling semver v1.0.22
   Compiling rustc-demangle v0.1.23
   Compiling anyhow v1.0.81
   Compiling either v1.10.0
   Compiling itoa v1.0.11
   Compiling ryu v1.0.17
   Compiling backtrace v0.3.71
   Compiling psm v0.1.21
   Compiling object v0.32.2
   Compiling stacker v0.1.15
   Compiling serde_json v1.0.115
   Compiling equivalent v1.0.1
   Compiling minimal-lexical v0.2.1
   Compiling csv-core v0.1.11
   Compiling anstyle v1.0.6
   Compiling anstyle-query v1.0.2
   Compiling anstyle-parse v0.2.3
   Compiling unicode_categories v0.1.1
   Compiling yansi v0.5.1
   Compiling unsafe-libyaml v0.2.11
   Compiling colorchoice v1.0.0
   Compiling miniz_oxide v0.7.2
   Compiling itertools v0.12.1
   Compiling ahash v0.8.11
   Compiling ariadne v0.4.0
   Compiling anstream v0.6.13
   Compiling nom v7.1.3
   Compiling hashbrown v0.14.3
   Compiling quote v1.0.35
   Compiling syn v2.0.58
   Compiling indexmap v2.2.6
   Compiling chumsky v0.9.3
   Compiling sqlformat v0.2.3
   Compiling addr2line v0.21.0
   Compiling serde_derive v1.0.197
   Compiling strum_macros v0.26.2
   Compiling enum-as-inner v0.6.0
   Compiling strum v0.26.2
   Compiling csv v1.3.0
   Compiling sqlparser v0.43.1
   Compiling serde_yaml v0.9.34+deprecated
   Compiling prqlc-ast v0.11.3
   Compiling prqlc-parser v0.11.3
   Compiling prqlc v0.11.3
   Compiling _ch_rust_prql v0.1.0 (/mnt/ch/ClickHouse/build_new_rust/rust/workspace/prql)
    Finished `release` profile [optimized] target(s) in 14.96s
[4/7] cd /mnt/ch/ClickHouse/build_new_rust/rust/workspace/pco && /u...profile=release --offline --locked -- -Cdefault-linker-libraries=no
   Unpacking better_io v0.1.0 (registry `/mnt/ch/ClickHouse/rust/../contrib/rust_vendor`)
   Unpacking pco v0.2.3 (registry `/mnt/ch/ClickHouse/rust/../contrib/rust_vendor`)
   Unpacking rand_core v0.6.4 (registry `/mnt/ch/ClickHouse/rust/../contrib/rust_vendor`)
   Unpacking rand_xoshiro v0.6.0 (registry `/mnt/ch/ClickHouse/rust/../contrib/rust_vendor`)
   Compiling rand_core v0.6.4
   Compiling better_io v0.1.0
   Compiling rand_xoshiro v0.6.0
   Compiling pco v0.2.3
   Compiling _ch_rust_pco v0.1.0 (/mnt/ch/ClickHouse/build_new_rust/rust/workspace/pco)
    Finished `release` profile [optimized] target(s) in 1.19s
[7/7] Running utility command for _cargo-build__ch_rust_pco
```

Note that this means that you will need to update the rust_vendor submodule with the changes:
* Fork the submodule
* Create a new commit and branch with the changes
* Change the submodule in CH to point to your fork and commit
* Once the CH PR is working, we can create a PR to the submodule to get the official changes there and then switch back the submodule to the official rust_vendor url.

rgolovachev commented:
When I run the following command:
> cargo local-registry --git --sync Cargo.lock "$CH_TOP_DIR"/contrib/rust_vendor

I encounter this problem:
```
$ cargo local-registry --git --sync Cargo.lock "$CH_TOP_DIR"/contrib/rust_vendor
error: no such command: `local-registry`

	View all installed commands with `cargo --list`
	Find a package to install `local-registry` with `cargo search cargo-local-registry`
```

I tried to install cargo-local-registry, but I ran into a problem again:

```
$ cargo install --version 0.2.6 cargo-local-registry
    Updating crates.io index
  Installing cargo-local-registry v0.2.6
    Updating crates.io index
   Compiling libc v0.2.153
   Compiling proc-macro2 v1.0.81
   Compiling unicode-ident v1.0.12
   Compiling cfg-if v1.0.0
   Compiling once_cell v1.19.0
   Compiling pkg-config v0.3.30
   Compiling memchr v2.7.2
   Compiling vcpkg v0.2.15
   Compiling autocfg v1.2.0
   Compiling regex-syntax v0.8.3
   Compiling thiserror v1.0.59
   Compiling version_check v0.9.4
   Compiling parking_lot_core v0.9.9
   Compiling crc32fast v1.4.0
   Compiling smallvec v1.13.2
   Compiling scopeguard v1.2.0
   Compiling crossbeam-utils v0.8.19
   Compiling hex v0.4.3
   Compiling typenum v1.17.0
   Compiling adler v1.0.2
   Compiling zeroize v1.7.0
   Compiling same-file v1.0.6
   Compiling itoa v1.0.11
   Compiling num-conv v0.1.0
   Compiling sha1_smol v1.0.0
   Compiling time-core v0.1.2
   Compiling powerfmt v0.2.0
   Compiling bytes v1.6.0
   Compiling num_threads v0.1.7
   Compiling subtle v2.5.0
   Compiling tinyvec_macros v0.1.1
   Compiling miniz_oxide v0.7.2
   Compiling walkdir v2.5.0
   Compiling const-oid v0.9.6
   Compiling time-macros v0.2.18
   Compiling minimal-lexical v0.2.1
   Compiling tinyvec v1.6.0
   Compiling serde v1.0.198
   Compiling bitflags v1.3.2
   Compiling remove_dir_all v0.5.3
   Compiling deranged v0.3.11
   Compiling fastrand v1.9.0
   Compiling signal-hook v0.3.17
   Compiling percent-encoding v2.3.1
   Compiling unicode-bidi v0.3.15
   Compiling base64ct v1.6.0
   Compiling log v0.4.21
   Compiling hashbrown v0.13.2
   Compiling fastrand v2.0.2
   Compiling curl v0.4.46
   Compiling generic-array v0.14.7
   Compiling form_urlencoded v1.2.1
   Compiling ahash v0.8.11
   Compiling lock_api v0.4.11
   Compiling num-traits v0.2.18
   Compiling openssl-probe v0.1.5
   Compiling hashbrown v0.12.3
   Compiling base16ct v0.2.0
   Compiling zerocopy v0.7.32
   Compiling pem-rfc7468 v0.7.0
   Compiling utf8parse v0.2.1
   Compiling anyhow v1.0.82
   Compiling home v0.5.9
   Compiling serde_json v1.0.116
   Compiling cpufeatures v0.2.12
   Compiling rustix v0.38.34
error[E0658]: use of unstable library feature 'proc_macro_byte_character'
   --> /home/rrgolovachev/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.81/src/wrapper.rs:871:21
    |
871 |                     proc_macro::Literal::byte_character(byte)
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #115268 <https://github.com/rust-lang/rust/issues/115268> for more information
    = help: add `#![feature(proc_macro_byte_character)]` to the crate attributes to enable
    = note: this compiler was built on 2024-03-31; consider upgrading it if it is out of date

error[E0658]: use of unstable library feature 'proc_macro_c_str_literals'
   --> /home/rrgolovachev/.cargo/registry/src/index.crates.io-6f17d22bba15001f/proc-macro2-1.0.81/src/wrapper.rs:898:21
    |
898 |                     proc_macro::Literal::c_string(string)
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #119750 <https://github.com/rust-lang/rust/issues/119750> for more information
    = help: add `#![feature(proc_macro_c_str_literals)]` to the crate attributes to enable
    = note: this compiler was built on 2024-03-31; consider upgrading it if it is out of date

   Compiling ryu v1.0.17
   Compiling aho-corasick v1.1.3
   Compiling nom v7.1.3
   Compiling der v0.7.9
   Compiling unicode-bom v1.1.4
   Compiling syn v1.0.109
   Compiling anstyle-parse v0.2.3
   Compiling linux-raw-sys v0.4.13
For more information about this error, try `rustc --explain E0658`.
error: could not compile `proc-macro2` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
error: failed to compile `cargo-local-registry v0.2.6`, intermediate artifacts can be found at `/var/tmp/cargo-installz7JhwU`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
```

After that `cargo local-registry --git --sync Cargo.lock "$CH_TOP_DIR"/contrib/rust_vendor` is still failing

rgolovachev commented:
My rust toolchain:

```
$ rustup toolchain list
nightly-2024-04-01-x86_64-unknown-linux-gnu (default)
```

Algunenano commented:
> I tried to install cargo-local-registry, but I ran into a problem again:

Change to the stable build, build the tool, change back to unstable.

```
$ rustup toolchain install stable
$ rustup toolchain default stable
$ cargo install --version 0.2.6 cargo-local-registry
$ rustup default nightly-2024-04-01
```

rgolovachev commented:
Finally, I have built CH with my codec locally. @Algunenano, thank you so much for the help!

I will deal with [rust_vendor submodule](https://github.com/ClickHouse/rust_vendor) a little bit later. I hope I won't have any big problems with this.

rgolovachev committed:
try to fix compilation

rgolovachev committed:
try to fix compilation #2

rgolovachev committed:
try to fix compilation 3

rgolovachev committed:
disable quantile codec in fasttest

rgolovachev commented:
I have changed the submodule in CH to point to my fork of rust_vendor, but CI still fails. Maybe I need the "submodule changed" label in order to fix it?

Felixoid commented:
The label is set automatically in case of changed submodules. But you haven't updated it; see the changed files. Nothing in `contrib/` is updated.

You should check out a different commit in the `contrib/module-name`, get back to the main repository, and commit changed submodule.

rgolovachev committed:
update contrib/rust_vendor

rgolovachev committed:
revert .gitmodules

rgolovachev committed:
Merge branch 'add-quantile-codec' of github.com:rgolovachev/ClickHouse into add-quantile-codec

rschu1ze commented on src/Compression/CompressionCodecQuantile.cpp:
```diff
@@ -0,0 +1,217 @@
+
+class CompressionCodecQuantile : public ICompressionCodec
+{
+public:
```
Could we mark the codec experimental for now (`ICodec::isExperimental()`)?

rgolovachev replied:
done

rgolovachev committed:
mark quantile codec as experimental

rgolovachev committed:
yet another try to fix CI

rgolovachev commented:
> How does it fare in performance and compression ratio?

I have done some measurements on [NY taxi](https://clickhouse.com/docs/en/getting-started/example-datasets/nyc-taxi) dataset. For comparison, I chose default codec, quantile (my) codec, zstd and gorilla codecs. The overall size of dataset is 1.63 GB (20mln rows). I have executed [this query](https://pastila.nl/?0203928b/0f00121f43c2f8930c1dc66aac9e7587#q5Ic1eosxule/5jCpGi3Qg==) in order to see compression rate. Here are the results:

Default codec:
```

   ┌─disk_name─┬─compressed─┬─uncompressed─┬─compr_rate─┬─────rows─┬─part_count─┐
1. │ default   │ 772.58 MiB │ 1.35 GiB     │       1.78 │ 20000000 │          5 │
   └───────────┴────────────┴──────────────┴────────────┴──────────┴────────────┘

```

Quantile codec:
```

   ┌─disk_name─┬─compressed─┬─uncompressed─┬─compr_rate─┬─────rows─┬─part_count─┐
1. │ default   │ 550.07 MiB │ 1.35 GiB     │       2.51 │ 20000000 │          5 │
   └───────────┴────────────┴──────────────┴────────────┴──────────┴────────────┘
  ```
  
  zstd codec:
  ```
  
   ┌─disk_name─┬─compressed─┬─uncompressed─┬─compr_rate─┬─────rows─┬─part_count─┐
1. │ default   │ 589.32 MiB │ 1.35 GiB     │       2.34 │ 20000000 │          5 │
   └───────────┴────────────┴──────────────┴────────────┴──────────┴────────────┘
```

Gorilla codec:
```

   ┌─disk_name─┬─compressed─┬─uncompressed─┬─compr_rate─┬─────rows─┬─part_count─┐
1. │ default   │ 937.94 MiB │ 1.35 GiB     │       1.47 │ 20000000 │          5 │
   └───────────┴────────────┴──────────────┴────────────┴──────────┴────────────┘

```

For performance measurements I used [this query](https://pastila.nl/?0160df6a/eb049dd7c1b7f7c0987c0f60241a521c#/nu4Wixidp6IUatd2hXw6w==). Here are the results:

Default codec:
```
Queries executed: 100.

localhost:9000, queries: 100, QPS: 9.757, RPS: 195142540.738, MiB/s: 4774.705, result RPS: 4878.564, result MiB/s: 0.093.

0.000%		0.086 sec.
10.000%		0.092 sec.
20.000%		0.094 sec.
30.000%		0.095 sec.
40.000%		0.098 sec.
50.000%		0.099 sec.
60.000%		0.100 sec.
70.000%		0.102 sec.
80.000%		0.103 sec.
90.000%		0.106 sec.
95.000%		0.108 sec.
99.000%		0.112 sec.
99.900%		0.119 sec.
99.990%		0.119 sec.
```

Quantile codec:
```
Queries executed: 100.

localhost:9000, queries: 100, QPS: 9.191, RPS: 183820096.504, MiB/s: 4497.670, result RPS: 4595.502, result MiB/s: 0.088.

0.000%		0.095 sec.
10.000%		0.100 sec.
20.000%		0.101 sec.
30.000%		0.102 sec.
40.000%		0.103 sec.
50.000%		0.104 sec.
60.000%		0.105 sec.
70.000%		0.107 sec.
80.000%		0.108 sec.
90.000%		0.111 sec.
95.000%		0.112 sec.
99.000%		0.122 sec.
99.900%		0.124 sec.
99.990%		0.124 sec.
```

zstd codec:
```
Queries executed: 100.

localhost:9000, queries: 100, QPS: 9.779, RPS: 195570734.080, MiB/s: 4785.182, result RPS: 4889.268, result MiB/s: 0.093.

0.000%		0.085 sec.
10.000%		0.091 sec.
20.000%		0.094 sec.
30.000%		0.095 sec.
40.000%		0.097 sec.
50.000%		0.098 sec.
60.000%		0.100 sec.
70.000%		0.101 sec.
80.000%		0.102 sec.
90.000%		0.106 sec.
95.000%		0.107 sec.
99.000%		0.110 sec.
99.900%		0.157 sec.
99.990%		0.157 sec.
```

Gorilla:
```
Queries executed: 100.

localhost:9000, queries: 100, QPS: 8.746, RPS: 174919771.147, MiB/s: 4279.899, result RPS: 4372.994, result MiB/s: 0.083.

0.000%		0.100 sec.
10.000%		0.103 sec.
20.000%		0.105 sec.
30.000%		0.106 sec.
40.000%		0.108 sec.
50.000%		0.109 sec.
60.000%		0.111 sec.
70.000%		0.113 sec.
80.000%		0.116 sec.
90.000%		0.120 sec.
95.000%		0.123 sec.
99.000%		0.132 sec.
99.900%		0.163 sec.
99.990%		0.163 sec.
```

rschu1ze commented:
@rgolovachev Interesting, thanks. Just for my understanding: does pcodec have a knob to toggle the maximum loss of precision (e.g. 5%, 10%, etc.)? Related to that, is it possible to calculate/obtain the loss of precision for pcodec-compressed floats, e.g. in the NY taxi data set?

rgolovachev commented:
> @rgolovachev Interesting, thanks. Just for my understanding: does pcodec have a knob to toggle the maximum loss of precision (e.g. 5%, 10%, etc.)? Related to that, is it possible to calculate/obtain the loss of precision for pcodec-compressed floats, e.g. in the NY taxi data set?

AFAIK, the pcodec does not have this feature and it positions itself as a lossless codec

rgolovachev committed:
change .gitmodules and update pco version

rgolovachev committed:
cosmetics

mwlon commented on src/Compression/CompressionCodecQuantile.cpp:
```diff
@@ -0,0 +1,218 @@
+namespace DB
+{
+
+class CompressionCodecQuantile : public ICompressionCodec
```
Why is this referred to as `Quantile`? Quantile Compression was the predecessor to Pcodec, and I think we should be clear that we're using Pcodec here.

mwlon commented on src/Compression/CompressionCodecQuantile.cpp:
```diff
@@ -0,0 +1,218 @@
+{
+    if (!WhichDataType(column_type).isFloat())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Quantile codec is not applicable for {} because the data type is not float",
```
Why limit to floats? Pcodec handles signed and unsigned ints too.

rgolovachev replied:
Initially, the task was to add this codec for FP data types: [58394](https://github.com/ClickHouse/ClickHouse/issues/58394), but I also see no reason not to add handling for integers. I will add ints handling a little bit later.

mwlon commented:
One you release an experiment codec, are you allowed to change the format for it later on? If not, we should probably think about the standalone vs. wrapped choice here a bit.

rschu1ze commented:
Yes, while marked experimental, a codec may change its format. But I guess it would be simple to write another version field at the beginning of the compressed stream (`doCompressData`). When version older than the current version is encountered during read, we could throw an exception instead of trying to read the old format with new code.

mwlon commented on rust/workspace/pco/src/lib.rs:
```diff
@@ -0,0 +1,208 @@
+  len: u32,
+  file_size: *mut u32
+) -> PcoError {
+  match pco::standalone::guarantee::file_size::<T::L>(len as usize, &pco::PagingSpec::EqualPagesUpTo(len as usize)) {
```
By using the default settings, the paging spec is implicitly `PagingSpec::EqualPagesUpTo(DEFAULT_MAX_PAGE_N)`

mwlon commented:
If you're compressing >10k numbers at a time, and the only ops you'll ever want to support are bulk decompression and maybe reading the first n entries, then standalone should work. It's versioned and gives a compatibility error during decompression if the file's version is higher.

If you might want to group together multiple pages (or what are we compressing here? I haven't delved into ClickHouse to understand yet) during compression, but still want seeking/random access/interleaved statistics, then you should eventually use wrapped.

But since you can change the format later, this is just an FYI.

rgolovachev committed:
add ints handling + rename quantile to pcodec (except for the file names)

rgolovachev committed:
rename files

rgolovachev committed:
try to fix style

rgolovachev committed:
fix

rgolovachev committed:
format test

rgolovachev committed:
try to fix perftests

rgolovachev commented:
I added integers handling and also renamed Quantile codec to Pcodec, but performance tests failed for some reason, here is the codecs_int_insert-err.log from the test output: https://pastila.nl/?00078b76/c147353d37dd18c6c82d491a114b7b4f#Ft/lLt6EJuLlpxm9bKt0eg==

However I was able to run performance tests locally and here are the results:

- [codecs_float_select](https://pastila.nl/?00356e13/9bb841c20dce455bf8e03d8567c4b306#uhjdjG6mchKA6dR6NtJl6g==)
- [codecs_float_insert](https://pastila.nl/?000a4161/626547081ce25aeb29ee0382f8138ca5#tcrTe38ejIOBMtEm2CXK+w==)
- [codecs_int_select](https://pastila.nl/?0002eaa7/f31fb36f9ffc4717afe11b9ede9b37e6#PQMCm/JFOix3tE5mRUpxPA==)
- [codecs_int_insert](https://pastila.nl/?00345855/94570a296f7db498a9059815191d31a8#fiXMRmHdKqDsipDO/FIw2g==)

alexey-milovidov commented:
Yes, the performance test runs the old and new server versions simultaneously and compares results. 
The old version does not have this feature, which is why it cannot run. We should just ignore this error in the pull requests, and performance tests will run after it is merged.

rgolovachev commented:
> * [codecs_float_select](https://pastila.nl/?00356e13/9bb841c20dce455bf8e03d8567c4b306#uhjdjG6mchKA6dR6NtJl6g==)
> * [codecs_float_insert](https://pastila.nl/?000a4161/626547081ce25aeb29ee0382f8138ca5#tcrTe38ejIOBMtEm2CXK+w==)
> * [codecs_int_select](https://pastila.nl/?0002eaa7/f31fb36f9ffc4717afe11b9ede9b37e6#PQMCm/JFOix3tE5mRUpxPA==)
> * [codecs_int_insert](https://pastila.nl/?00345855/94570a296f7db498a9059815191d31a8#fiXMRmHdKqDsipDO/FIw2g==)

As you can see from these logs, there are significant latencies in insert queries. I tried to figure out why this is happening. I have started clickhouse server and `perf top` connected to the server. In clickhouse client I created table codec_mon_Float64_Pcodec (the same as in the perftest) and inserted numbers:

```INSERT INTO codec_mon_Float64_Pcodec (n) SELECT number+sin(number) FROM system.numbers LIMIT 20000000 SETTINGS max_threads=1```

So here are the results from perf tool: https://pastila.nl/?00d2b5a4/0c5e87eb1e6e801368c40f47d2ea1353#fHnfztGMTOlzPLdEgUzbyQ==

As you can see, `pco::bin_optimization::optimize_bins` spends a lot of CPU resources. In this regard, I have a question. @mwlon, do you have any idea why this function might be spending so many CPU resources and are there any ways to reduce this consumption?

mwlon commented:
This makes me think it's doing a lot of small compressions. Is it not just compressing all 20M numbers at once? If it's compressing <100k at a time, that could be a real performance issue. If it's <10k at a time, it's probably a compression ratio problem too. Normally pco should get 100's of MB/s, even on an old CPU core, but this seems to be more like 30.

Also, this isn't the code path you're hitting, but you might want to update pcodec. There was a compression speedup last week.

Things to try:

* Can we increase the page size (or whatever it's called) to confirm that fixes it?
* If that's the problem, we can use the wrapped format to group multiple pages together together. 
* If we can't use the wrapped format, we can at least update and decrease compression level, which will make it tolerably fast. But that's a last resort.

rgolovachev commented:
Just updating the library to version 0.2.5 didn't help. But by inserting `std::cerr` at the beginning of the `doCompressData` method execution, I found out that we always compress 8192 numbers simultaneously. I tried to increase the number of simultaneously compressible numbers by increasing `min_compress_block_size` and `max_compress_block_size` from `src/Core/Settings.h` and increasing `DEFAULT_BLOCK_SIZE` along with `DEFAULT_INSERT_BLOCK_SIZE` from `src/Core/Defines.h`. After these changes, `doCompressData` compressed 131072 numbers in about half of the cases. And it really helped! Here are the results from the insert perftests:

- [codecs_float_insert](https://pastila.nl/?000d24cd/e8b65d0bc606955def8a73e52413ce56#sKjmokEEbJP2eVHKG6JsRg==)
- [codecs_int_insert](https://pastila.nl/?000b175c/9acc85bf6e3dc8402def3c6899bae580#R7oz1gWSB2ej70dK/e7dTw==)

So the small size of the data that needs to be compressed was really a problem. So we need to think about whether we can use the wrapped format.

mwlon commented:
@rgolovachev Thanks, so that confirms it's the small chunk size. What is the structure of the clickhouse internal format? I can't seem to find any docs about it. Ideally here's how we'd implement it:

During compression, we take a sizeable amount of data (ideally ~100k-2M numbers, somewhat flexible) and make a new `pco::wrapped::ChunkCompressor`, telling it the pages sizes we want (apparently 8192). This does the bulk of compression work, but doesn't write anything out. Then we write its chunk metadata somewhere[1] in the file, and we can write out each page individually. These pages shared the chunk metadata and require it to decompress, akin to Zstd dictionaries. A decompressor must first read the chunk metadata, and then it will be able to decompress any page individually.

[1] Some formats weren't designed with the flexibility to write metadata/dictionaries shared by multiple pages. I hope that's not the case for Clickhouse, but if it is, we'll have to settle for the version bump and reduced compression level I suggested. You can model the compression time as `a * 4^compression_level + b * n * compression_level + c`, and for tiny chunks like these at default compression level the `a * 4^compression_level` term dominates.

rgolovachev commented:
Sorry for abandoning the PR, but I haven't come up with a good way to share chunk metadata between multiple compressed data blocks. So I had to lower the compression level and I was choosing between level 4 and 5. I was looking at the perftests results and compression rate on the NY taxi dataset. In addition, @alexey-milovidov advised trying different codecs + zstd, so here are the results:

pcodec with default level 4:
- [codecs_float_insert](https://pastila.nl/?005eacfa/283047da1c7166c2632c16e4ba39e536#KVH9rav0LBN4rK6sFb+7BQ==)
- [codecs_int_insert](https://pastila.nl/?00181084/251a1667e612b6e004e6a75d69e2932f#5Etqf9MpjLU4uW8nQphZXw==)

pcodec with default level 5:
- [codecs_float_insert](https://pastila.nl/?006b8c12/753153d3cbcb905656306f9b9dee4259#NCG5A4DRiPSPObfyTvGvtw==)
- [codecs_int_insert](https://pastila.nl/?000881b3/fb1477b6ae352d8803119f756cfa3627#iaS2Hsw3Q8ovNFbpca1LqQ==)

In my opinion compression level 4 would be preferable.

Compression rate on the NY taxi dataset for different codecs: [link](https://pastila.nl/?01411d5f/4a6cd9dbeb340951b11bb612b95bac0f#EaUVJ/XGaRhazW9HutSmAQ==)

So as you can see, using pcodec together with zstd makes sense.

mwlon commented:
Thanks @rgolovachev . If I'm reading this right, looks like Pco level 4 is a bit faster than the other codecs compared and also compressed a bit better? Sounds good. So far we haven't considered small chunks of numbers much, so it's possible we'll be able to improve performance even more in the future.

Did you also measure the compression speed for Pco+zstd? Pco is designed to be used by itself and doesn't byte-align much of anything, so using them together is a bit cursed, but I'm not opposed to it if you like the characteristics.

rgolovachev commented:
> If I'm reading this right, looks like Pco level 4 is a bit faster than the other codecs compared and also compressed a bit better?

pco level 4 is often much faster than the other codecs on SELECT queries, but slightly inferior on INSERT queries. On the NY taxi dataset, pco with level 4 compresses data slightly better than zstd. But you can offer some more datasets for pco compression rate comparison.

> Did you also measure the compression speed for Pco+zstd?

Here I manually executed the queries from the codecs_float_insert perftest: [pcodec compression level 4 + zstd](https://pastila.nl/?00152f3c/b98c3d4ca4bc71460da5525c182eeeb2#Hj0X36lhtbvdEGWTbhvCJg==), [pcodec compression level 5 + zstd](https://pastila.nl/?01a885c7/766a64500f915526b78a9573b2cef76c#Qc08trP3Hr1qCS/3Et/nmg==)

You can see that the performance has almost not degraded (compared to the pcodec results in the perftests without zstd)

rschu1ze commented:
This PR stalled.

We already have FPC and Gorilla as lossless floating-point codecs and soon we will get SZ3 as a lossy codec #67161. The aforementioned codecs are all written in C++ whereas pcodec is Rust. Also, the aforementioned codecs are well established in the sense that people benchmarked them systematically and wrote papers about them (see https://github.com/mwlon/pcodec/issues/189), so users know what can be expected from them. I don't see similar evaluations for pcodec (maybe it needs some time to mature).

I'll close the PR, and if there are good arguments we can reopen in future.
